packages/errors/lib/base-error.js

```diff
@@ -0,0 +1,170 @@
+	 * A machine-readable error code which identifies the specific type of error.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {string}
+	 */
+	code = BaseError.defaultCode;
+
+	/**
+	 * The root cause error instance.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {Error | null}
+	 */
+	cause = null;
+
+	/**
+	 * Additional error information.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {{[key: string]: any}}
+	 */
+	data = {};
+
+	/**
+	 * Create an error with no arguments.
+	 *
+	 * @overload
+	 */
+	/**
+	 * Create an error with error data.
+	 *
+	 * @overload
+	 * @param {ErrorData} data
+	 *     Additional error information.
+	 */
+	/**
+	 * Create an error with a message and optional error data.
+	 *
+	 * @overload
+	 * @param {string} message
+	 *     The error message.
+	 * @param {ErrorData} [data]
+	 *     Additional error information.
+	 */
+	/**
+	 * @param {string | ErrorData} [message]
+	 *     The error message if it's a string, or full error information if an object.
+	 * @param {ErrorData} [data]
+	 *     Additional error information if `message` is a string.
+	 */
+	constructor(message, data = {}) {
+		if (typeof message === 'string') {
+			data.message = message;
+		} else {
+			data = message || data;
+		}
+		super(data.message || BaseError.defaultMessage);
+
+		if (typeof data.code === 'string') {
+			this.code = BaseError.normalizeErrorCode(data.code);
+		}
+
+		if (data.cause instanceof Error) {
+			this.cause = data.cause;
+		}
+
+		for (const [key, value] of Object.entries(data)) {
+			const self = /** @type {typeof BaseError} */ (this.constructor);
+			if (!self.reservedKeys.includes(key)) {
+				this.data[key] = value;
+			}
+		}
+	}
+
+	/**
+	 * Reserved keys that should not appear in `BaseError.prototype.data`.
+	 *
+	 * @protected
+	 * @type {string[]}
+	 */
+	static reservedKeys = ['code', 'message', 'cause'];
+
+	/**
+	 * @protected
+	 * @readonly
+	 * @type {string}
+	 */
+	static defaultCode = 'UNKNOWN';
+
+	/**
+	 * @protected
+	 * @readonly
+	 * @type {string}
+	 */
+	static defaultMessage = 'An error occurred';
+
+	/**
+	 * Get whether an error object is marked as operational (it has a truthy `isOperational` property).
+	 *
+	 * @public
+	 * @param {Error} error
+	 *     The error object to check.
+	 * @returns {boolean}
+	 *     Returns whether the error is operational.
+	 */
+	static isErrorMarkedAsOperational(error) {
+		// @ts-ignore Error.prototype.isOperational does not exist, but it's OK to check in this
+		// case as we're manually casting `undefined` to a Boolean
+		return Boolean(error.isOperational);
+	}
+
+	/**
+	 * Normalize a machine-readable error code.
+	 *
+	 * @private
+	 * @param {string} code
+	 *     The error code to normalize.
+	 * @returns {string}
+	 *     Returns the normalized error code.
+	 */
+	static normalizeErrorCode(code) {
+		return code
+			.trim()
+			.toUpperCase()
+			.replace(/[^a-z0-9_]+/gi, '_');
+	}
+}
+
+module.exports = BaseError;
+
+module.exports.default = module.exports;
```

Comment:
We're doing this in all the Reliability Kit modules for compatibility with the different module systems we use. In certain cases a TypeScript project will compile this:

```js
import BaseError from '@dotcom-reliability-kit/errors/lib/base-error';
```

to the following CommonJS:

```js
const BaseError = require('@dotcom-reliability-kit/errors/lib/base-error').default;
```

We don't really have any control over this, and aliasing the `default` export like this helps us avoid having to compile the code as part of a build step.

Reply:
Oh and in regular CommonJS land, we don't want people to have to specify `default`, so we change `module.exports` to allow users to do this:

```js
const BaseError = require('@dotcom-reliability-kit/errors/lib/base-error');
```

rather than having to remember to add `.default` to the end.

Reply:
Ah I seeee, ok. That helps to understand. Thank you 😸

Reply:
So for me to bear in mind, if we're building something (e.g. a new package) that may be used by different module systems, it's preferred to double export like this to ensure it's compatible once installed?